### DISCUSSION TITLE
Time series: connect null value setting ignored when showing nearest point tooltip

### DISCUSSION THREAD

**Discussion by Knocks83** [grafana/grafana#38284] [Dashboards, Panels & Visualizations]
**What happened**:
I somehow (sorry, I can't remember how I enabled that setting) enabled a setting that makes the graph show the tooltip of the nearest point. Now, even if I set "connect null values" as never, it still connects the points.
![image](https://user-images.githubusercontent.com/22529815/119110769-bb5caf00-ba22-11eb-9390-57d99cd9c48f.png)
It also happens outside the edit view

**What you expected to happen**:
I expected the points to not connect to each other

**How to reproduce it (as minimally and precisely as possible)**:
Set a Time series panel to show the tooltip of the nearest point, set "connect null values" to never.

**Anything else we need to know?**:
If you could tell me how to enable the setting that shows the nearest point's tooltip you'd do me a favor :P

**Environment**:
- Grafana version: 8.0.0 beta 2
- Data source type & version: InfluxDB 1.8.5
- OS Grafana is installed on: Ubuntu 20.04
- User OS & Browser: Windows 10 - Firefox

**Comment by leeoniya** [comment]
i don't think there's any setting related to tooltips that affects null connected-ness. is this the same issue as https://github.com/grafana/grafana/issues/34166? and was that issue fixed by 8 Beta 2 for you?

if the connected issue still exists, then please provide the actual raw data you're rendering, since i am unable to reproduce this. you can grab the raw query response from the Network tab in Chrome's Devtools:

![image](https://user-images.githubusercontent.com/43234/119157177-864c5e80-ba1a-11eb-804e-c61128736f2a.png)

you may have multiple query requests, so make sure you grab them all. make sure to remove any sensitive data before sharing. thanks!

**Comment by Knocks83** [comment]
@leeoniya that fix partaially fixed it: the dual-query graph got fixed, while that single-query graph still has that problem.
That's the query:
```
{
    "results": [
        {
            "statement_id": 0,
            "series": [
                {
                    "name": "environment",
                    "columns": [
                        "time",
                        "status"
                    ],
                    "values": [
                        [
                            1621590464666,
                            false
                        ],
                        [
                            1621590647994,
                            false
                        ],
                        [
                            1621590831262,
                            false
                        ],
                        [
                            1621591014611,
                            false
                        ],
                        [
                            1621591197938,
                            false
                        ],
                        [
                            1621591381310,
                            false
                        ],
                        [
                            1621591564674,
                            false
                        ],
                        [
                            1621591749244,
                            false
                        ],
                        [
                            1621591932557,
                            false
                        ],
                        [
                            1621592115904,
                            false
                        ],
                        [
                            1621592299228,
                            false
                        ],
                        [
                            1621592482559,
                            false
                        ],
                        [
                            1621592665877,
                            false
                        ],
                        [
                            1621592849341,
                            false
                        ],
                        [
                            1621593032679,
                            false
                        ],
                        [
                            1621593215998,
                            false
                        ],
                        [
                            1621593399325,
                            false
                        ],
                        [
                            1621593582682,
                            false
                        ],
                        [
                            1621593765944,
                            false
                        ],
                        [
                            1621593949283,
                            false
                        ],
                        [
                            1621594132869,
                            false
                        ],
                        [
                            1621594316232,
                            false
                        ],
                        [
                            1621594499727,
                            false
                        ],
                        [
                            1621594682995,
                            false
                        ],
                        [
                            1621594866389,
                            false
                        ],
                        [
                            1621595049702,
                            false
                        ],
                        [
                            1621595233088,
                            false
                        ],
                        [
                            1621595417408,
                            false
                        ],
                        [
                            1621595600727,
                            false
                        ],
                        [
                            1621595784064,
                            false
                        ],
                        [
                            1621595967420,
                            false
                        ],
                        [
                            1621596150790,
                            false
                        ],
                        [
                            1621596334160,
                            false
                        ],
                        [
                            1621596517431,
                            false
                        ],
                        [
                            1621596700702,
                            false
                        ],
                        [
                            1621596884129,
                            false
                        ],
                        [
                            1621597067444,
                            false
                        ],
                        [
                            1621597250802,
                            false
                        ],
                        [
                            1621597434096,
                            false
                        ],
                        [
                            1621597617446,
                            false
                        ],
                        [
                            1621597800848,
                            false
                        ],
                        [
                            1621597984210,
                            false
                        ],
                        [
                            1621598167474,
                            false
                        ],
                        [
                            1621598350790,
                            false
                        ],
                        [
                            1621598534132,
                            false
                        ],
                        [
                            1621598717399,
                            false
                        ],
                        [
                            1621598900717,
                            false
                        ],
                        [
                            1621599085101,
                            false
                        ],
                        [
                            1621599268370,
                            false
                        ],
                        [
                            1621599451675,
                            false
                        ],
                        [
                            1621599634972,
                            false
                        ],
                        [
                            1621599818293,
                            false
                        ],
                        [
                            1621600001602,
                            false
                        ],
                        [
                            1621600184901,
                            false
                        ],
                        [
                            1621600368245,
                            false
                        ],
                        [
                            1621600551512,
                            false
                        ],
                        [
                            1621600734869,
                            false
                        ],
                        [
                            1621600918209,
                            false
                        ],
                        [
                            1621601101540,
                            false
                        ],
                        [
                            1621601284844,
                            false
                        ],
                        [
                            1621601468324,
                            false
                        ],
                        [
                            1621601651612,
                            false
                        ],
                        [
                            1621601835080,
                            false
                        ],
                        [
                            1621602018380,
                            false
                        ],
                        [
                            1621602201859,
                            false
                        ],
                        [
                            1621602385154,
                            false
                        ],
                        [
                            1621602568510,
                            false
                        ],
                        [
                            1621602753076,
                            false
                        ],
                        [
                            1621602936436,
                            false
                        ],
                        [
                            1621603119764,
                            false
                        ],
                        [
                            1621603303098,
                            false
                        ],
                        [
                            1621603486494,
                            false
                        ],
                        [
                            1621603669802,
                            false
                        ],
                        [
                            1621603853110,
                            false
                        ],
                        [
                            1621604036457,
                            false
                        ],
                        [
                            1621604219875,
                            false
                        ],
                        [
                            1621604403163,
                            false
                        ],
                        [
                            1621604586496,
                            false
                        ],
                        [
                            1621604769856,
                            false
                        ],
                        [
                            1621604953196,
                            false
                        ],
                        [
                            1621605136576,
                            false
                        ],
                        [
                            1621605319927,
                            false
                        ],
                        [
                            1621605503269,
                            false
                        ],
                        [
                            1621605686670,
                            false
                        ],
                        [
                            1621605870007,
                            false
                        ],
                        [
                            1621606053545,
                            false
                        ],
                        [
                            1621606236923,
                            false
                        ],
                        [
                            1621606431463,
                            false
                        ],
                        [
                            1621606614848,
                            false
                        ],
                        [
                            1621606798172,
                            false
                        ],
                        [
                            1621606981525,
                            false
                        ],
                        [
                            1621607164870,
                            false
                        ],
                        [
                            1621607348182,
                            false
                        ],
                        [
                            1621607531496,
                            false
                        ],
                        [
                            1621607714893,
                            false
                        ],
                        [
                            1621607898235,
                            false
                        ],
                        [
                            1621608082378,
                            false
                        ],
                        [
                            1621608265669,
                            false
                        ],
                        [
                            1621608449046,
                            false
                        ],
                        [
                            1621608632445,
                            false
                        ],
                        [
                            1621608815769,
                            false
                        ],
                        [
                            1621608999100,
                            false
                        ],
                        [
                            1621609182400,
                            false
                        ],
                        [
                            1621609365742,
                            false
                        ],
                        [
                            1621609549102,
                            false
                        ],
                        [
                            1621609732410,
                            false
                        ],
                        [
                            1621609915709,
                            false
                        ],
                        [
                            1621610100442,
                            false
                        ],
                        [
                            1621610283735,
                            false
                        ],
                        [
                            1621610467045,
                            false
                        ],
                        [
                            1621610650370,
                            false
                        ],
                        [
                            1621610833645,
                            false
                        ],
                        [
                            1621611017027,
                            false
                        ],
                        [
                            1621611200320,
                            false
                        ],
                        [
                            1621611383590,
                            false
                        ],
                        [
                            1621611566912,
                            false
                        ],
                        [
                            1621611750246,
                            false
                        ],
                        [
                            1621611933513,
                            false
                        ]
                    ]
                }
            ]
        }
    ]
}

```

This one is the panel JSON:
```
{
  "id": 4,
  "gridPos": {
    "h": 3,
    "w": 24,
    "x": 0,
    "y": 19
  },
  "type": "timeseries",
  "pluginVersion": "8.0.0-beta1",
  "fieldConfig": {
    "defaults": {
      "custom": {
        "drawStyle": "line",
        "lineInterpolation": "stepBefore",
        "barAlignment": 0,
        "lineWidth": 1,
        "fillOpacity": 10,
        "gradientMode": "none",
        "spanNulls": 600000,
        "showPoints": "never",
        "pointSize": 5,
        "stacking": {
          "mode": "none",
          "group": "A"
        },
        "axisPlacement": "auto",
        "axisLabel": "Status",
        "scaleDistribution": {
          "type": "linear"
        },
        "hideFrom": {
          "tooltip": false,
          "graph": false,
          "legend": false
        },
        "thresholdsStyle": {
          "mode": "off"
        }
      },
      "color": {
        "mode": "palette-classic"
      },
      "thresholds": {
        "mode": "absolute",
        "steps": [
          {
            "color": "green",
            "value": null
          }
        ]
      },
      "mappings": [],
      "decimals": 0,
      "max": 1,
      "min": 0,
      "unit": "none"
    },
    "overrides": []
  },
  "options": {
    "tooltip": {
      "mode": "single"
    },
    "legend": {
      "displayMode": "hidden",
      "placement": "bottom",
      "calcs": []
    }
  },
  "targets": [
    {
      "alias": "Status",
      "groupBy": [],
      "measurement": "environment",
      "orderByTime": "ASC",
      "policy": "default",
      "refId": "A",
      "resultFormat": "time_series",
      "select": [
        [
          {
            "params": [
              "status"
            ],
            "type": "field"
          }
        ]
      ],
      "tags": []
    }
  ],
  "timeFrom": null,
  "timeShift": null,
  "transparent": true,
  "datasource": null
}
```

Also, I'm attaching a video of that setting (that looks like it doesn't exist) because I can actually see a different behavior between the 2 graphs.
As you can see, in the graph on top I have to manually find a point to see the tooltips values, while the one on the bottom always shows the nearest point's value.

https://user-images.githubusercontent.com/22529815/119165821-32fbff80-ba5e-11eb-8338-efd2fc629d7a.mp4

**Comment by Knocks83** [comment]
Nevermind, that happens with new panels too. If you want, I can give you the server address in DM

**Comment by leeoniya** [comment]
let me see if i can explain this situation in a way that will make sense :)

for better value alignment i will use `_` in place of `null` and small numbers instead of full timestamps in the snippets below.

let's take a simple dataset with 1 series and no null values:

```
1, 3, 5, 7     // timestamps (x)
9, 9, 9, 9     // temp values (y1)
```

when you move your cursor to x = 3.2, the chart will try to find the closest x value in the data, which will be 3. the hover points and tooltip will then be shown for all datapoints at 3. your tooltip will look like this:

```
Date: 3     // shared x
-------
Temp: 9     // y1 value at x=3
```

now let's say you have a null value:

```
1, 3, 5, 7     // timestamps (x)
9, _, 9, 9     // temp values (y1)
```

in this case, when you hover 3.2, you will **not** see a hover point for temp, nor a tooltip because at the closest x value (3), the y value for temp is `null`:

https://user-images.githubusercontent.com/43234/119242300-61431300-bb22-11eb-96ca-1c7e5a556306.mp4

you can see that there is no extra behavior to tries to scan left or right for a non-null temp value. it found the closest x offset, and simply read the data from all y series at that x position.

if we add another series with additional timestamps:

```
1, 3, 6, 8     // timestamps (x)
0, 0, 0, 0     // rh% values (y1)
```

this dataset is then "joined" (similar to SQL JOIN) to the other dataset:

```
1, 3, 5, 6, 7, 8     // timestamps (x)
9, _, 9, _, 9, _     // temp values (y1)
0, 0, _, 0, _, 0     // rh% values (y2)
```

the behavior is the same. where the values are null (either explicitly or as a result of the join), you will not see hover points or tooltip values, because for that closest X, the y values in those series are not recorded.

https://user-images.githubusercontent.com/43234/119242481-b3d0ff00-bb23-11eb-9f76-89f068606eb1.mp4

whether the the chart is configured to visually connects across null values has no effect on this, because the underlying data is the same.

you can see that we cannot simply scan to the nearest non-null value in each series, since the tooltip has a single shared X display, and it would become quite inaccurate to show the values of each series at wildly different timestamps while showing a single timestamp in the tooltip header. while this may be okay-ish for dense data, it is not okay when one series has datapoints a few hours apart and another has values seconds apart.

so, what you're showing happening for a shared cursor follows the above rules.

1. the cursors are synchronized by time along x
2. each chart finds the closest x index in its data to that cursor position
3. it displays a tooltip containing non-null values for all series at that index.

since the data in your top chart is dense, and presumably has a lot of null values, and the lower chart is more sparse, it gives the impression that it works differently in one chart (by looking for the closest non-null value).

for a chart with a single series, you can "fix" this by removing all `null` values from the dataset, but this will not work when more than 1 series have to be joined, since it inserts nulls automatically into places that are un-aligned. (you can try splitting your temp and humidity panel into two panels, which will avoid the join)

in https://github.com/grafana/grafana/pull/34552 i have added a patch that actually scans for non-null data in each series to focus the points at different timestamps (the tooltip is not fixed yet). while this can improve expected behavior in some cases, it has the opposite effect of being very confusing in other cases where the data is misaligned and sparse between series and charts:

https://user-images.githubusercontent.com/43234/119242735-0e6b5a80-bb26-11eb-9fb0-2a9c7946c103.mp4

this means we have to have some "reasonable" or configurable threshold for how far we should scan left and right, and have to figure out how to display multiple timestamps instead of one in a tooltip to prevent misinterpretation of the vis.

the UX issues are trickier than the technical ones.

**Comment by Knocks83** [comment]
@leeoniya thanks for your kind explaination! It sounded strange at the beginning since, just to make a comparison, I only have 1 table with timestamp, temp, humidity and status, so there shouldn't have been any different behavior between the graphs...
Well, there was a null filling in the graph on the top, so that was that problem.

The bug is another one tho, what about the graph on the bottom that connects when it shouldn't? #34166 fixed the dual-query graph, but the single-query graph still has tha problem. Are we reopening that issue?

By removing the null-filling and the group by time $__interval (so removing all null points) that's what I noticed:
![image](https://user-images.githubusercontent.com/22529815/119253865-70be6c80-bbb3-11eb-9d3c-3be3d5448cf5.png)

Even the graph that worked before, now isn't working anymore! So the issue is probably related to the absence of null-values(?)


edit:
I tried to fill with none instead of null (so keeping time($__interval)) and it still gives me that problem

**Comment by leeoniya** [comment]
> Even the graph that worked before, now isn't working anymore! So the issue is probably related to the absence of null-values(?)

this is probably right. the only way to display gaps in the chart is to have `null` values in your data. when you select "connect null values", the pre-processing will _remove_ null values from your data, which will cause the graph to connect. "always" will remove all null values, "threshold" will remove null values from only specific time regions. there is nothing that will insert gaps into your data that were not previously there -- perhaps this is what you were expecting threshold to do?

please provide the query response(s) for the chart you are showing. it is very difficult to help you without it. you have only provided a query response with all false values (the bottom chart), which is not very useful.

**Comment by Knocks83** [comment]
> there is nothing that will insert gaps into your data that were not previously there -- perhaps this is what you were expecting threshold to do?

That's right, it's what I was expecting. So right now I have to choose between:

1. Having null values, being able to "split" the graph depending on the treshold but being unable to see the nearest point's values
2. Not having null values, being unable to "split" the graph but being able to see the nearest point's values

Right? Would it be a reasonable feature request to ask for a way to do both? (with reasonable I mean do-able) Also considering that I cannot aggregate boolean values, so I cannot null-fill the bottom graph and so I cannot "split" the bottom graph.

**Comment by leeoniya** [comment]
> Would it be a reasonable feature request to ask for a way to do both?

yes, it may even be reasonable for this to be default behavior in threshold mode. we'll discuss this internally in the next few days.

**Comment by oscarkilhed** [comment]
I'm closing this for now. We can let any future feature request refer back to this.

**Comment by leeoniya** [comment]
i've landed https://github.com/grafana/grafana/pull/34552

please try the next nightly build or next beta release:

![image](https://user-images.githubusercontent.com/43234/126842180-027834f8-dbe0-4dd3-bd8d-b23aabc96119.png)

**Comment by Knocks83** [comment]
@leeoniya Just updated to 8.1.10-beta2 but they still connect even if the treshold is smaller than the difference between the time of the 2 points.
In my case I have 7 days of difference between the points and have set a 10m treshold
![image](https://user-images.githubusercontent.com/22529815/126864610-c349c6a3-6e7a-42cb-a07c-adabb7623261.png)

As per usual, that's the panel JSON:
```json
{
  "id": 2,
  "gridPos": {
    "h": 19,
    "w": 24,
    "x": 0,
    "y": 0
  },
  "type": "timeseries",
  "title": "Environment",
  "datasource": "Localhost",
  "pluginVersion": "8.0.0-beta1",
  "fieldConfig": {
    "defaults": {
      "custom": {
        "drawStyle": "line",
        "lineInterpolation": "linear",
        "barAlignment": 0,
        "lineWidth": 1,
        "fillOpacity": 10,
        "gradientMode": "none",
        "spanNulls": 600000,
        "showPoints": "never",
        "pointSize": 5,
        "stacking": {
          "mode": "none",
          "group": "A"
        },
        "axisPlacement": "auto",
        "axisLabel": "Temperature",
        "scaleDistribution": {
          "type": "linear"
        },
        "hideFrom": {
          "tooltip": false,
          "viz": false,
          "legend": false,
          "graph": false
        },
        "thresholdsStyle": {
          "mode": "off"
        },
        "axisWidth": 60,
        "lineStyle": {
          "fill": "solid"
        }
      },
      "color": {
        "mode": "palette-classic"
      },
      "thresholds": {
        "mode": "absolute",
        "steps": [
          {
            "color": "green",
            "value": null
          }
        ]
      },
      "mappings": [],
      "decimals": 1,
      "max": 40,
      "min": 10,
      "unit": "celsius"
    },
    "overrides": [
      {
        "matcher": {
          "id": "byName",
          "options": "Humidity"
        },
        "properties": [
          {
            "id": "color",
            "value": {
              "fixedColor": "blue",
              "mode": "fixed"
            }
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "Temperature"
        },
        "properties": [
          {
            "id": "color",
            "value": {
              "fixedColor": "red",
              "mode": "fixed"
            }
          }
        ]
      },
      {
        "matcher": {
          "id": "byName",
          "options": "Humidity"
        },
        "properties": [
          {
            "id": "unit",
            "value": "percent"
          },
          {
            "id": "max",
            "value": 100
          },
          {
            "id": "custom.axisLabel",
            "value": "Humidity"
          }
        ]
      }
    ]
  },
  "options": {
    "tooltip": {
      "mode": "multi"
    },
    "legend": {
      "displayMode": "list",
      "placement": "bottom",
      "calcs": [
        "lastNotNull",
        "max",
        "min"
      ]
    }
  },
  "targets": [
    {
      "alias": "Temperature",
      "groupBy": [
        {
          "params": [
            "$__interval"
          ],
          "type": "time"
        },
        {
          "params": [
            "none"
          ],
          "type": "fill"
        }
      ],
      "hide": false,
      "measurement": "environment",
      "orderByTime": "ASC",
      "policy": "default",
      "query": "SELECT mean(\"temperature\") FROM \"environment\" WHERE $timeFilter GROUP BY time($__interval) fill(null)",
      "rawQuery": false,
      "refId": "A",
      "resultFormat": "time_series",
      "select": [
        [
          {
            "params": [
              "temperature"
            ],
            "type": "field"
          },
          {
            "params": [],
            "type": "mean"
          }
        ]
      ],
      "tags": []
    },
    {
      "alias": "Humidity",
      "groupBy": [
        {
          "params": [
            "$__interval"
          ],
          "type": "time"
        },
        {
          "params": [
            "none"
          ],
          "type": "fill"
        }
      ],
      "measurement": "environment",
      "orderByTime": "ASC",
      "policy": "default",
      "query": "SELECT mean(\"value\") FROM \"Humidity\" WHERE $timeFilter GROUP BY time($__interval) fill(null)",
      "rawQuery": false,
      "refId": "B",
      "resultFormat": "time_series",
      "select": [
        [
          {
            "params": [
              "humidity"
            ],
            "type": "field"
          },
          {
            "params": [],
            "type": "mean"
          }
        ]
      ],
      "tags": []
    }
  ],
  "transparent": true,
  "timeFrom": null,
  "timeShift": null
}
```

I'm also attaching the CSV data of one of the 2 queries in the graph, just in case you need it.
[Data.zip](https://github.com/grafana/grafana/files/6872204/Data.zip)

**Comment by leeoniya** [comment]
@Knocks83 the setting you're showing is called "Connect null values < threshold"; it is not "disconnect adjacent values > threshold", which would be a new feature request. the dataset you attached does not contain null values, so they are all connected.

if you insert a null value in that time range, then it will disconnect:

![image](https://user-images.githubusercontent.com/43234/126873318-b626bb50-befd-4e12-b28e-5424d1b1efa9.png)

my previous comment was referring to the improved tooltip behavior for mis-aligned series, which is the video you posted in https://github.com/grafana/grafana/issues/34542#issuecomment-846062363, where you are unable to hover nearby datapoints without very precise cursor positioning. the connectedness of adjacent datapoints appears to work as intended. the tooltip improvements are not in beta2 since they only landed yesterday, but should be available in the nightly build already or beta3 (which is due out next week, i think).

**Comment by Knocks83** [comment]
@leeoniya yeah sorry, I just installed the latest nightly and it contains your pr. It works! Thanks!

**Comment by leeoniya** [comment]
i added an API to uPlot that allows arbitrary gap insertion during drawing, which does not require `null` insertion in the data.

https://leeoniya.github.io/uPlot/demos/missing-data.html (bottom chart)

whether we end up exposing control over this in Grafana remains to be decided.

**Comment by Knocks83** [comment]
> i added an API to uPlot that allows arbitrary gap insertion during drawing, which does not require `null` insertion in the data.

It looks sick! I know that won't be used by a lot of people, since null-filling is unavailable only for specific data types (boolean in my case), but thanks a lot for your work!

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by natellium** [comment]
Hey there! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

I will be closing this discussion, but if you’re still interested in this request being implemented please don’t hesitate to open a [feature request](https://github.com/grafana/grafana/issues/new?assignees=&labels=&projects=&template=1-feature_requests.md). Sorry for the inconvenience, and thanks for your contributions!

cc @grafana/dataviz-squad
